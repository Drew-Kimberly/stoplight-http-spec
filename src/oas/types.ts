import {
  IApiKeySecurityScheme,
  IBasicSecurityScheme,
  IBearerSecurityScheme,
  IOauth2SecurityScheme,
  IOpenIdConnectSecurityScheme,
} from '@stoplight/types';
import { OpenAPIObject } from 'openapi3-ts';
import { Spec } from 'swagger-schema-official';

export declare type HttpSecurityScheme =
  | IApiKeySecurityScheme
  | IBearerSecurityScheme
  | IBasicSecurityScheme
  | IOauth2SecurityScheme
  | IOpenIdConnectSecurityScheme;

// prefixing with T to avoid tslinter "no-shadow" error
export interface ITranslateOperationOptions<TSpec> {
  document: TSpec;
  path: string;
  method: string;
}

export type Oas2TranslateOperationOptions = ITranslateOperationOptions<Spec>;
export type Oas3TranslateOperationOptions = ITranslateOperationOptions<OpenAPIObject>;

export const enum OasVersion {
  OAS2 = 2,
  OAS3 = 3,
}
