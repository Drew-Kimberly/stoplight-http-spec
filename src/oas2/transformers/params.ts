import {
  HttpParamStyles,
  IHttpContent,
  IHttpEncoding,
  IHttpHeaderParam,
  IHttpOperationRequestBody,
  IHttpPathParam,
  IHttpQueryParam,
} from '@stoplight/types';
import { defaults, get, map, pick, pickBy } from 'lodash';
import {
  BodyParameter,
  FormDataParameter,
  Header,
  HeaderParameter,
  PathParameter,
  QueryParameter,
  Schema,
} from 'swagger-schema-official';

function chooseQueryParameterStyle(
  parameter: QueryParameter,
): HttpParamStyles.PipeDelimited | HttpParamStyles.SpaceDelimited | HttpParamStyles.Form {
  /** Must cast to 'any' because this field is missing from the types but it's defined in the spec
   * https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#parameterObject
   */
  switch (parameter.collectionFormat) {
    case 'pipes':
      return HttpParamStyles.PipeDelimited;
    case 'ssv':
      return HttpParamStyles.SpaceDelimited;
    case 'csv':
    case 'multi':
    default:
      /**
       * This implementation is, in fact, not fully compliant with oas3.
       * As per oas3 spec: "Form style parameters defined by RFC6570.
       *                    This option replaces collectionFormat with a csv
       *                    (when explode is false) or multi (when explode is true)
       *                    value from OpenAPI 2.0."
       * But since there is no such property like 'explode' in oas2 we are defaulting to 'form'.
       */
      return HttpParamStyles.Form;
  }
}

export function translateToHeaderParam(parameter: HeaderParameter): IHttpHeaderParam {
  return (pickBy({
    ...buildContentFromParameter(parameter),
    name: parameter.name,
    style: HttpParamStyles.Simple,
    required: parameter.required,
  }) as unknown) as IHttpHeaderParam;
}

export function translateToHeaderParams(headers: { [headerName: string]: Header }): IHttpHeaderParam[] {
  return map(headers, (header, name) => {
    const param: IHttpHeaderParam = {
      ...buildContentFromParameter(Object.assign({ name }, header)),
      name,
      style: HttpParamStyles.Simple,
    };
    return param;
  });
}

export function translateToBodyParameter(body: BodyParameter, consumes: string[]): IHttpOperationRequestBody {
  const examples = map(get(body, 'x-examples'), (value, key) => ({ key, value }));

  return pickBy({
    description: body.description,
    required: body.required,
    contents: consumes.map(mediaType => {
      return {
        mediaType,
        schema: body.schema,
        examples,
      };
    }),
  });
}

export function translateFromFormDataParameter(
  form: FormDataParameter,
  requestBody: IHttpOperationRequestBody | null | undefined,
  consumes: string[],
): IHttpOperationRequestBody {
  const baseContent = buildContentFromParameter(form);

  const requestBodyCopy: IHttpOperationRequestBody = Object.assign(
    {},
    requestBody || {
      allowEmptyValue: form.allowEmptyValue,
      contents: consumes.map(mediaType => ({
        ...baseContent,
        mediaType,
      })),
    },
  );
  const bodyContent = requestBodyCopy.contents && requestBodyCopy.contents[0];

  if (bodyContent) {
    bodyContent.schema = bodyContent.schema || baseContent.schema;

    const encoding: IHttpEncoding = {
      property: form.name,
      style: HttpParamStyles.Form,
    };

    bodyContent.encodings = (bodyContent.encodings || []).concat(encoding);
  }

  return requestBodyCopy;
}

export function translateToQueryParameter(query: QueryParameter): IHttpQueryParam {
  return (pickBy({
    ...buildContentFromParameter(query),
    allowEmptyValue: query.allowEmptyValue,
    name: query.name,
    style: chooseQueryParameterStyle(query),
    required: query.required,
  }) as unknown) as IHttpQueryParam;
}

export function translateToPathParameter(parameter: PathParameter): IHttpPathParam {
  return (pickBy({
    ...buildContentFromParameter(parameter),
    name: parameter.name,
    style: HttpParamStyles.Simple,
    required: parameter.required,
  }) as unknown) as IHttpPathParam;
}

function buildContentFromParameter(
  param: QueryParameter | PathParameter | HeaderParameter | FormDataParameter | Header,
): IHttpContent {
  const schema: Schema = pick(
    param,
    'type',
    'format',
    'default',
    'description',
    'enum',
    'exclusiveMaximum',
    'exclusiveMinimum',
    'maxItems',
    'maxLength',
    'maximum',
    'minItems',
    'minimum',
    'minLength',
    'title',
    'items',
    'pattern',
    'uniqueItems',
    'multipleOf',
  );

  return {
    schema: pickBy(
      defaults<Schema, Schema>(schema, {
        minLength: 'allowEmptyValue' in param && param.allowEmptyValue ? 1 : undefined,
      }),
    ),
  };
}
