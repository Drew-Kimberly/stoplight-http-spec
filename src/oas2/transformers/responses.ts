import { IHttpOperationResponse, IMediaTypeContent } from '@stoplight/types';
import { JSONSchema4 } from 'json-schema';
import { map, partial } from 'lodash';
import { Response } from 'swagger-schema-official';

import { translateToHeaderParams } from './params';

const toObject = <T>(value: T, key: string) => ({ key, value });

function translateToResponse(produces: string[], response: Response, statusCode: string): IHttpOperationResponse {
  const headers = translateToHeaderParams(response.headers || {});
  const objectifiedExamples = map(response.examples, toObject);

  const contents: IMediaTypeContent[] = [];

  produces.forEach(produceElement => {
    if (response.examples && response.examples[produceElement]) {
      contents.push({
        mediaType: produceElement,
        schema: response.schema as JSONSchema4,
        examples: objectifiedExamples.filter(example => example.key === produceElement),
      });
    }
  });

  const translatedResponses = {
    code: statusCode,
    description: response.description,
    headers,
    contents,
  };

  if (translatedResponses.contents.length === 0)
    translatedResponses.contents[0] = {
      mediaType: '',
      schema: {},
      examples: [],
    };

  translatedResponses.contents[0].examples!.push(
    ...objectifiedExamples.filter(example => !produces.includes(example.key)),
  );

  return translatedResponses;
}

export function translateToResponses(
  responses: { [name: string]: Response },
  produces: string[],
): IHttpOperationResponse[] & { 0: IHttpOperationResponse } {
  return map(responses, partial(translateToResponse, produces)) as IHttpOperationResponse[] & {
    0: IHttpOperationResponse;
  };
}
