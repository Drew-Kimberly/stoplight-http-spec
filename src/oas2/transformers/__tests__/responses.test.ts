import { HttpParamStyles, IHttpHeaderParam } from '@stoplight/types';

import { translateToHeaderParams } from '../params';
import { translateToResponses } from '../responses';

jest.mock('../params');

describe('responses', () => {
  const fakeHeaderParams: IHttpHeaderParam[] = [{ name: 'fake-header', style: HttpParamStyles.Simple }];
  const produces = ['application/json', 'application/xml'];

  beforeEach(() => {
    (translateToHeaderParams as jest.Mock).mockReturnValue(fakeHeaderParams);
  });

  test('should translate to multiple responses', () => {
    const responses = translateToResponses(
      {
        r1: {
          description: 'd1',
          examples: {
            'application/json': {},
          },
          headers: {},
          schema: {},
        },
        r2: {
          description: 'd2',
          examples: {
            'application/xml': {},
          },
          headers: {},
          schema: {},
        },
      },
      produces,
    );

    expect(responses).toMatchSnapshot();
  });

  test('should translate to response w/o headers', () => {
    expect(
      translateToResponses(
        {
          r1: {
            description: 'd1',
            examples: {
              'application/xml': {},
            },
            schema: {},
          },
        },
        produces,
      ),
    ).toMatchSnapshot();
  });

  test('should translate to response w/o examples', () => {
    expect(
      translateToResponses(
        {
          r1: {
            description: 'd1',
            schema: {},
          },
        },
        produces,
      ),
    ).toMatchSnapshot();
  });

  describe('should keep foreign examples', () => {

    it('by creating a new content if necessary', () => {

      const responses = translateToResponses(
        {
          r1: {
            description: 'd1',
            examples: {
              'application/i-have-no-clue': {},
            },
            headers: {},
            schema: {},
          },
        },
        produces,
      );

      expect(responses[0].contents).toBeDefined();
      expect(responses[0].contents![0].examples).toBeDefined();
      expect(responses[0].contents![0].examples![0]).toBeDefined();
      expect(responses[0].contents![0].examples![0]).toHaveProperty('key', 'application/i-have-no-clue')
    })
    it('aggregating to the first example', () => {
      const responses = translateToResponses(
        {
          r1: {
            description: 'd1',
            examples: {
              'application/i-have-no-clue': {},
              'application/json': {}
            },
            headers: {},
            schema: {},
          },
        },
        produces,
      );

      expect(responses[0].contents).toBeDefined();
      expect(responses[0].contents![0].examples).toBeDefined();
      expect(responses[0].contents![0].examples![0]).toBeDefined();
      expect(responses[0].contents![0].examples![0]).toHaveProperty('key', 'application/json')
    })
  });
});
