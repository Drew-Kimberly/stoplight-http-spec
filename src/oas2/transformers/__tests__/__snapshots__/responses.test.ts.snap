// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`responses should translate to multiple responses 1`] = `
Array [
  Object {
    "code": "r1",
    "contents": Array [
      Object {
        "examples": Array [
          Object {
            "key": "application/json",
            "value": Object {},
          },
        ],
        "mediaType": "application/json",
        "schema": Object {},
      },
    ],
    "description": "d1",
    "headers": Array [
      Object {
        "name": "fake-header",
        "style": "simple",
      },
    ],
  },
  Object {
    "code": "r2",
    "contents": Array [
      Object {
        "examples": Array [
          Object {
            "key": "application/xml",
            "value": Object {},
          },
        ],
        "mediaType": "application/xml",
        "schema": Object {},
      },
    ],
    "description": "d2",
    "headers": Array [
      Object {
        "name": "fake-header",
        "style": "simple",
      },
    ],
  },
]
`;

exports[`responses should translate to response w/o examples 1`] = `
Array [
  Object {
    "code": "r1",
    "contents": Array [
      Object {
        "examples": Array [],
        "mediaType": "",
        "schema": Object {},
      },
    ],
    "description": "d1",
    "headers": Array [
      Object {
        "name": "fake-header",
        "style": "simple",
      },
    ],
  },
]
`;

exports[`responses should translate to response w/o headers 1`] = `
Array [
  Object {
    "code": "r1",
    "contents": Array [
      Object {
        "examples": Array [
          Object {
            "key": "application/xml",
            "value": Object {},
          },
        ],
        "mediaType": "application/xml",
        "schema": Object {},
      },
    ],
    "description": "d1",
    "headers": Array [
      Object {
        "name": "fake-header",
        "style": "simple",
      },
    ],
  },
]
`;
